src/System.Web.OData/OData/Query/Expressions/PropertyContainer.cs

```diff
@@ -54,29 +56,44 @@ protected PropertyContainer()
         //          Name = properties[0].Key, 
         //          Value = properties[0].Value,
         //
-        //          Next = new NamedProperty<> { ..... } 
+        //          LeftNext = new NamedProperty<> { ..... } 
+        //          RightNext = new NamedProperty<> { ..... } 
         //      }
         public static Expression CreatePropertyContainer(IList<NamedPropertyExpression> properties)
         {
             Expression container = null;
 
             // build the linked list of properties.
-            foreach (NamedPropertyExpression property in properties)
+            if (properties.Count >= 1)
             {
-                container = CreateNamedPropertyCreationExpression(property, container);
+                NamedPropertyExpression property = properties.First();
+                int count = properties.Count - 1;
+                int leftSize = GetLeftSize(count);
+                Expression leftNext = CreatePropertyContainer(properties.Skip(1).Take(leftSize).ToList());
```

Comment:
With recursion we will have to call ToArray() on ArraySegment, because you can't create ArryaSegment on top of ArraySegment and let it deal with offsets  => leads to the same number of allocations, but makes code more comlex

---
In reply to: [131730512](https://github.com/kosinsky/WebApi/pull/9#discussion_r131730512) [](ancestors = 131730512)